library/modules/Buildings.cpp

```diff
@@ -1376,61 +1307,130 @@ static void on_civzone_delete(df::building_civzonest* civzone)
+            // already queued for destruction
+            return false;
+        }
         bld->queueDestroy();
         return false;
     }
 
-    /* Immediate destruction code path.
-       Should only happen for abstract and unconstructed buildings.*/
+    // Immediate destruction code path.
+    // Should only happen for abstract and unconstructed buildings.
 
-    if (bld->isSettingOccupancy())
-    {
+    if (bld->flags.bits.almost_deleted)
+        return true;
+    bld->flags.bits.almost_deleted = true;
+
+    if (bld->isSettingOccupancy()) {
         markBuildingTiles(bld, true);
         bld->cleanupMap();
     }
 
-    bld->removeUses(false, false);
-    // Assume: no parties.
-    unlinkRooms(bld);
-    // Assume: not unit destroy target
-    int id = bld->id;
-    vector_erase_at(plotinfo->tax_collection.rooms, linear_index(plotinfo->tax_collection.rooms, id));
-    // Assume: not used in punishment
-    // Assume: not used in non-own jobs
-    // Assume: does not affect pathfinding
-    bld->deconstructItems(false, false, false);
-    // Don't clear arrows.
-
-    bld->uncategorize();
+    bool destitem = bld->isFarmPlot();
+    bool insanify = false;
+    bool is_from_damage = false;
 
-    remove_building_from_all_zones(bld);
-
-    if (bld->getType() == df::building_type::Civzone)
-    {
-        auto zone = strict_virtual_cast<df::building_civzonest>(bld);
+    bld->removeUses(destitem, insanify);
 
-        if (zone)
+    const df::building_type btype = bld->getType();
+    if (btype == df::building_type::Civzone) {
+        if (auto zone = strict_virtual_cast<df::building_civzonest>(bld))
             on_civzone_delete(zone);
+    } else {
+        while (bld->relations.size())
+            remove_building_from_zone(bld, bld->relations[bld->relations.size() - 1]);
+    }
+
+    for (auto unit : world->units.active) {
+        if (unit->job.destroy_target == bld)
+            unit->job.destroy_target = NULL;
+    }
+
+    const int id = bld->id;
+    erase_from_vector(plotinfo->tax_collection.rooms, id);
+
+    for (auto punishment : plotinfo->punishments) {
+        if (punishment->chain == id)
+            punishment->chain = -1;
+    }
+
+    const bool interrupt_rest =
+        btype == df::building_type::TractionBench ||
+        btype == df::building_type::Bed;
+    const bool interrupt_link =
+        btype == df::building_type::Door ||
+        btype == df::building_type::Floodgate ||
+        btype == df::building_type::Bridge ||
+        btype == df::building_type::Support ||
+        btype == df::building_type::Chain ||
+        btype == df::building_type::Cage ||
+        btype == df::building_type::Weapon ||
+        btype == df::building_type::Hatch ||
+        btype == df::building_type::GrateWall ||
+        btype == df::building_type::GrateFloor ||
+        btype == df::building_type::BarsVertical ||
+        btype == df::building_type::BarsFloor ||
+        btype == df::building_type::GearAssembly;
```

Comment:
`df::building_type::Trap` with `df::trap_type::Lever`, `df::trap_type::PressurePlate`, or `df::trap_type::TrackStop`.

Reply:
yeah, I figured just checking for `df::building_type::Trap` would be sufficient here. I don't foresee any false positives or false negatives with the simpler check